lib/bolt/cli.rb

```diff
@@ -434,7 +434,7 @@ def execute(options)
             config.project
           )
 
-          plugins = Bolt::Plugin.setup(config, pal, analytics)
+          plugins = Bolt::Plugin.new(config, pal, analytics)
```

Comment:
So, this is a bit weird. `Plugin.setup` will resolve any plugins called under the `plugins` key in configuration so that any plugin defaults are already loaded when we go to resolve plugins. However, we already call `config_for_plugin` on plugins before resolving them to load the config, which will resolve the config `plugins` key, and I assume we do that for inventory plugins too.

This change only works because all plugins already call `config_for_plugin` themselves, which resolves `config.plugins`. By bypassing this we require any other configuration options that we make pluggable to also call `config_for_plugin` on their own, which could be confusing to figure out if we ever make other options pluggable. That's a big if though, since we haven't heard of anyone wanting to make other config options pluggable in *looks at watch* years, plus I can't imagine a situation where you wouldn't call `config_for_plugin` since that's how you get the plugin config to begin with.

If requiring any config options to call `config_for_plugin` on their own is what we want to do (which I feel pretty comfortable with), then it seems like we should remove `setup` altogether so that we don't *sometimes* load all plugins that are configured, and sometimes require the plugins to be resolved by calling `by_name` manually. It also seems like `setup` is just unnecessary at this point, we added it around having PuppetDB client config be configurable with plugins, and plugin loading has changed so much since hten. It seems like all `Plugin.setup` calls can be replaced with `Plugin.new`:
* In `BoltSpec::Context` this gets passed to the real Inventory, which resolves any plugins (if the above works, then the call in `BoltSpec::Context` should too)
* Same in `BoltSpec::Run`
* The empty inventory....might be fine? I initially thought it wouldn't matter but need to test if you have an empty inventory but then call `resolve_references` from a plan
* `BoltServer::TransportApp` also adds all their own modules manually

Reply:
I was initially going to drop usage of `setup` altogether, but wasn't sure of the ramifications elsewhere (particularly BoltServer). If you're comfortable dropping it then I can pick up those changes tomorrow.